### PR TITLE
Tests to  warmup_cosine_decay_schedule edge cases

### PR DESCRIPTION
**Pull request by edawite**
Added unit tests for warmup_cosine_decay_schedule covering edge cases, dtype handling, and JIT consistency to improve reliability of this common schedule. All tests pass locally: 6 passed.

### CODE REVIEW

**Comment by rdyro** [comment]
Can you explain the justification for adding these tests? We have some warmup cosine tests here already: https://github.com/google-deepmind/optax/blob/e5649e7805cf5c399bdf7e15a0521ce2cc372cd3/optax/schedules/_schedule_test.py#L484

Do you want to extend them? Are there any cases you're interested in testing in particular?

**Comment by edawite** [comment]
These tests were mainly to cover the edge case where decay_steps == warmup_steps, plus checking dtype handling and JIT/vmap consistency. I can fold them into _schedule_test.py if that’s the better place.

File: optax/schedules/schedules_warmup_cosine_test.py [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+        )
+
+        vb = _sample(base, decay_steps)
+        vs = _sample(steeper, decay_steps)
+
+        # After warmup, exponent=2 should never exceed exponent=1 pointwise.
+        for t in range(warmup_steps, decay_steps + 1):
+            self.assertLessEqual(float(vs[t]), float(vb[t]) + 1e-12)
+
+    def test_accepts_int_steps_and_returns_float_dtype(self):
```
Comment (rdyro) [inline optax/schedules/schedules_warmup_cosine_test.py]: I don't think this test is necessary

File: optax/schedules/schedules_warmup_cosine_test.py [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+            init_value=init,
+            peak_value=peak,
+            warmup_steps=warmup_steps,
+            decay_steps=decay_steps,
+            end_value=end,
+        )
+        sched_jit = jax.jit(sched)
+
+        steps = jnp.arange(decay_steps + 1)
+        eager = jax.vmap(sched)(steps)
```
Comment (rdyro) [inline optax/schedules/schedules_warmup_cosine_test.py]: I don't understand the vmap use here

File: optax/schedules/schedules_warmup_cosine_test.py [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+
+def _sample(sched, n_steps):
+    """Return values at steps [0, 1, ..., n_steps]."""
+    steps = jnp.arange(n_steps + 1)
+    return jax.vmap(sched)(steps)
+
+
+class WarmupCosineScheduleTest(parameterized.TestCase):
+
+    def test_warmup_increases_and_reaches_peak(self):
```
Comment (rdyro) [inline optax/schedules/schedules_warmup_cosine_test.py]: can you fold this into the next test?

**Review by rdyro — COMMENTED** [review]
I added some non-exhaustive, comments, could you compress these tests into fewer separate cases and fold them into existing schedule tests please?

**Comment by edawite** [comment]
Folded tests into _schedule_test.py and compressed into two cases as suggested and passed all tests locally.

**Comment by rdyro** [comment]
Are you proposing to remove existing tests?

Could you comment the test by surrounding code blocks with [subTest](https://github.com/google-deepmind/optax/blob/606a5d8e99a48ecacef1212c9c8cf1736ab2818c/optax/tree_utils/_casting_test.py#L64)?

**Comment by edawite** [comment]
Wrapped cases with subTest

File: optax/schedules/_schedule_test.py [hunk 4]
```diff
@@ -484,72 +484,102 @@ def test_with_giant_int_steps(self):
...
+          rtol=1e-6, atol=1e-8,
+      )
 
+  def test_raises_when_decay_equals_warmup(self):
+    with self.assertRaises(ValueError):
+      _schedule.warmup_cosine_decay_schedule(
+          init_value=0.2, peak_value=1.2,
+          warmup_steps=10, decay_steps=10, end_value=0.0)
 
+#ee
```
Comment (rdyro) [inline optax/schedules/_schedule_test.py]: remove this leftover comment

**Comment by edawite** [comment]
Restored all original tests unchanged.

**Comment by edawite** [comment]
Removed duplicated tests.

**Review by rdyro — APPROVED** [review]
